src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -15,27 +16,33 @@ public class CompilationWithAnalyzers
         private readonly Compilation _compilation;
         private readonly CancellationToken _cancellationToken;
         private readonly ConcurrentSet<Diagnostic> _exceptionDiagnostics;
+        private ImmutableHashSet<SyntaxTree> _analyzedSyntaxTrees = ImmutableHashSet<SyntaxTree>.Empty;
+        private Dictionary<SyntaxTree, Task<ImmutableArray<Diagnostic>>> _documentTasks = new Dictionary<SyntaxTree, Task<ImmutableArray<Diagnostic>>>();
+        private Task<ImmutableArray<Diagnostic>> _completeAnalysisTask;
+        private Task _latestAnalysisTask;
+        private readonly object _analysisLock = new object();
 
         public Compilation Compilation
         {
             get { return _compilation; }
         }
-
+        
         /// <summary>
         /// Creates a new compilation by attaching diagnostic analyzers to an existing compilation.
         /// </summary>
         /// <param name="compilation">The original compilation.</param>
         /// <param name="analyzers">The set of analyzers to include in future analyses.</param>
         /// <param name="options">Options that are passed to analyzers.</param>
+        /// <param name="onAnalyzerException">Action to invoke if an analyzer throws an exception.</param>
         /// <param name="cancellationToken">A cancellation token that can be used to abort analysis.</param>
-        public CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, CancellationToken cancellationToken)
+        public CompilationWithAnalyzers(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, AnalyzerOptions options, Action<Exception, DiagnosticAnalyzer, Diagnostic> onAnalyzerException, CancellationToken cancellationToken)
         {
             _cancellationToken = cancellationToken;
```

Comment:
An (unsatisfactory) short answer is that changing the API at this point requires a vigorous argument with the compatibility council. However, that is worth doing if the current API is demonstrably broken.

The intended model is that the cancellation token provided at construction can be used to cancel all analysis. That falls down when there is no central request for analysis, just a collection of clients. (In fact, that issue arises and is commented upon elsewhere in this change set.)

Cancelling some analysis of a single CompilationWithAnalyzers without cancelling all of its analysis is not supported. (That implies that the commented-upon usage elsewhere is not correct and must change.) This conclusion is not entirely comfortable.